Source/SwiftLintBuiltInRules/Rules/Style/TrailingClosureRule.swift

```diff
@@ -18,127 +19,81 @@ struct TrailingClosureRule: OptInRule {
             Example("offsets.sorted { $0.offset < $1.offset }"),
             Example("foo.something({ return 1 }())"),
             Example("foo.something({ return $0 }(1))"),
-            Example("foo.something(0, { return 1 }())")
+            Example("foo.something(0, { return 1 }())"),
+            Example("for x in list.filter({ $0.isValid }) {}"),
+            Example("if list.allSatisfy({ $0.isValid }) {}"),
+            Example("foo(param1: 1, param2: { _ in true }, param3: 0)"),
+            Example("foo(param1: 1, param2: { _ in true }) { $0 + 1 }"),
+            Example("foo(param1: { _ in false }, param2: { _ in true })"),
+            Example("foo(param1: { _ in false }, param2: { _ in true }, param3: { _ in false })"),
+            Example("""
+            if f({ true }), g({ true }) {
+                print("Hello")
+            }
+            """),
+            Example("""
+            for i in h({ [1,2,3] }) {
+                print(i)
+            }
+            """)
         ],
         triggeringExamples: [
             Example("↓foo.map({ $0 + 1 })"),
             Example("↓foo.reduce(0, combine: { $0 + 1 })"),
             Example("↓offsets.sorted(by: { $0.offset < $1.offset })"),
-            Example("↓foo.something(0, { $0 + 1 })")
+            Example("↓foo.something(0, { $0 + 1 })"),
+            Example("↓foo.something(param1: { _ in true }, param2: 0, param3: { _ in false })"),
+            Example("""
+            for n in list {
+                ↓n.forEach({ print($0) })
+            }
+            """)
         ]
     )
+}
 
-    func validate(file: SwiftLintFile) -> [StyleViolation] {
-        let dict = file.structureDictionary
-        return violationOffsets(for: dict, file: file).map {
-            StyleViolation(ruleDescription: Self.description,
-                           severity: configuration.severityConfiguration.severity,
-                           location: Location(file: file, byteOffset: $0))
-        }
-    }
-
-    private func violationOffsets(for dictionary: SourceKittenDictionary, file: SwiftLintFile) -> [ByteCount] {
-        var results = [ByteCount]()
+private extension TrailingClosureRule {
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override func visitPost(_ node: FunctionCallExprSyntax) {
+            guard node.trailingClosure == nil else { return }
 
-        if dictionary.expressionKind == .call,
-            shouldBeTrailingClosure(dictionary: dictionary, file: file),
-            let offset = dictionary.offset {
-            results = [offset]
-        }
+            if configuration.onlySingleMutedParameter {
+                guard node.containsOnlySingleMutedParameter else { return }
```

Comment:
Condition can be combined with the `if` above. So it's clearer that both conditions must hold.

Reply:
I'm not sure if I understand your point correctly. 

If we combine two conditions into one with `&&`, it makes sense for the case when both of them are `true` but it won't for the case when `configuration.onlySingleMutedParameter` is `true` and `node.containsOnlySingleMutedParameter` is `false` because it will incorrectly evaluate `else` or `else if` below and can append an incorrect violation when we want it to exit the function without adding any. Doing this causes `TrailingClosureRuleTests` to fail (at least on my machine).

Reply:
You are right. I confused myself thinking about the whole method and how it could be written differently.

Reply:
Fixed at f371ffb